specs/HubProtocol.md

```diff
@@ -0,0 +1,351 @@
+
+**NOTE**: `Invocation ID`s are arbitrarily chosen by the Caller and the Callee is expected to use the same string in all response messages. Callees may establish reasonable limits on `Invocation ID` lengths and terminate the connection when an `Invocation ID` that is too long is received.
+
+## Multiple Results
+
+The SignalR protocol allows for multiple `Result` messages to be transmitted in response to an `Invocation` message, and allows the receiver to dispatch these results as they arrive, to allow for streaming data from one endpoint to another.
+
+On the Callee side, it is up to the Callee's Binder to determine if a method call will yield multiple results. For example, in .NET certain return types may indicate multiple results, while others may indicate a single result. Even then, applications may wish for multiple results to be buffered and returned in a single `Completion` frame. It is up to the Binder to decide how to map this. The Callee's Binder must encode each result in separate `Result` messages, indicating the end of results by sending a `Completion` message. Since the `Completion` message accepts an optional payload value, methods with single results can be handled with a single `Completion` message, bearing the complete results.
+
+On the Caller side, the user code which performs the invocation indicates how it would like to receive the results and it is up the Caller's Binder to determine how to handle the result. If the Caller expects only a single result, but multiple results are returned, the Caller's Binder should yield an error indicating that multiple results were returned. However, if a Caller expects multiple results, but only a single result is returned, the Caller's Binder should yield that single result and indicate there are no further results.
+
+## Errors
+
+Errors are indicated by the presence of the `error` field in a `Completion` message. Errors always indicate the immediate end of the invocation. In the case of streamed responses, the arrival of a `Completion` message indicating an error should **not** stop the dispatching of previously-received results. The error is only yielded after the previously-received results have been dispatched.
+
+## Examples
+
+Consider the following C# methods
+
+```csharp
+public int Add(int x, int y)
+{
+    return x + y;
+}
+
+public int SingleResultFailure(int x, int y)
+{
+    throw new Exception("It didn't work!");
+}
+
+public IEnumerable<int> Batched(int count)
+{
+    for(var i = 0; i < count; i++)
+    {
+        yield return i;
+    }
+}
+
+[return: StreamToClient] // This is a made-up attribute that is used to indicate to the .NET Binder that it should stream results
```

Comment:
I'd prefer `Streamed`, since this is a declarative attribute on the return value. Also, we can nit-pick the attribute later, it's not really part of the spec.